acme/acme/crypto_util.py

```diff
@@ -359,6 +379,85 @@ def _pyopenssl_cert_or_req_san(cert_or_req: Union[crypto.X509, crypto.X509Req])
+
+
+def make_self_signed_cert(private_key: CertificateIssuerPrivateKeyTypes,
+                          domains: Optional[List[str]] = None,
+                          not_before: Optional[int] = None,
+                          validity: int = (7 * 24 * 60 * 60), force_san: bool = True,
+                          extensions: Optional[List[x509.Extension]] = None,
+                          ips: Optional[List[Union[ipaddress.IPv4Address,
+                                                   ipaddress.IPv6Address]]] = None
+                          ) -> x509.Certificate:
+    """Generate new self-signed certificate.
+    :param buffer private_key_pem: Private key, in PEM PKCS#8 format.
+    :type domains: `list` of `str`
+    :param int not_before: A POSIX UTC timestamp after which the cert is valid
+    :param validity: Time (in seconds) for which the cert will be valid
+    :param buffer private_key_pem: One of `CertificateIssuerPrivateKeyTypes`
+    :param bool force_san:
+    :param extensions: List of additional extensions to include in the cert.
+    :type extensions: `list` of `x509.Extension[x509.ExtensionType]`
+    :type ips: `list` of (`ipaddress.IPv4Address` or `ipaddress.IPv6Address`)
+    If more than one domain is provided, all of the domains are put into
+    ``subjectAltName`` X.509 extension and first domain is set as the
+    subject CN. If only one domain is provided no ``subjectAltName``
+    extension is used, unless `force_san` is ``True``.
+    """
+    assert domains or ips, "Must provide one or more hostnames or IPs for the cert."
+
+    builder = x509.CertificateBuilder()
+    builder = builder.serial_number(x509.random_serial_number())
+
+    if extensions is not None:
+        for ext in extensions:
+            builder = builder.add_extension(ext.value, ext.critical)
+    if domains is None:
+        domains = []
+    if ips is None:
+        ips = []
+    builder = builder.add_extension(x509.BasicConstraints(ca=True, path_length=0), critical=True)
+
+    name_attrs = []
+    if len(domains) > 0:
+        name_attrs.append(x509.NameAttribute(
+            x509.OID_COMMON_NAME,
+            domains[0]
+        ))
+
+    builder = builder.subject_name(x509.Name(name_attrs))
+    builder = builder.issuer_name(x509.Name(name_attrs))
+
+    sanlist: List[x509.GeneralName] = []
+    for address in domains:
+        sanlist.append(x509.DNSName(address))
+    for ip in ips:
+        sanlist.append(x509.IPAddress(ip))
+    if force_san or len(domains) > 1 or len(ips) > 0:
+        builder = builder.add_extension(
+            x509.SubjectAlternativeName(sanlist),
+            critical=False
+        )
+
+    not_before_datetime = datetime.fromtimestamp(
+        0 if not_before is None else not_before
+    )
+    builder = builder.not_valid_before(not_before_datetime)
+    validity_duration = timedelta(seconds=validity)
+    if not_before is None:
+        not_valid_after = _now() + validity_duration
```

Comment:
This doesn't match the docs -- this results in a certificate that's valid from (1970, now() + validity).

Reply:
ah i actually figured that was the behavior we'd want. i'm guessing it should default to `(1970, 1970 + validity)`?

Reply:
Sigh, I just realized there was an implicit misunderstanding about what the 0 was doing in the pyOpenSSL case:

The gmtime_adj methods are both relative to `now()`. So 0 there means "now", so it was actually the not_before time that was inconsistent.

As you can tell, the pyOpenSSL APIs are atrocious and this is why we want to burn it down :-)

Reply:
lol we love a confusing-yet-pervasive API, don't we folks. updated!